### PR TITLE
Add messages to indicate that opening an external firewall may be necessary.

### PR DESCRIPTION
**Pull request by brendandburns**
@thockin @juliaferraioli @jlowdermilk 

Closes https://github.com/GoogleCloudPlatform/kubernetes/issues/9150

### CODE REVIEW

File: pkg/kubectl/cmd/create.go [hunk 1]
```diff
@@ -113,3 +117,24 @@ func RunCreate(f *cmdutil.Factory, out io.Writer, filenames util.StringList) err
 	}
 	return nil
 }
+
+func printObjectSpecificMessage(obj runtime.Object, out io.Writer) {
+	switch obj := obj.(type) {
+	case *api.Service:
+		if obj.Spec.Type == api.ServiceTypeLoadBalancer {
```
Comment (thockin) [inline pkg/kubectl/cmd/create.go]: Also NodePort, although that probably should say "you might want to ..."

Comment (thockin) [inline pkg/kubectl/cmd/create.go]: Where do we document the overly-loose GCE firewall problem?  Should we stick it in a .md and link it here?

File: pkg/kubectl/cmd/create.go [hunk 2]
```diff
@@ -113,3 +117,24 @@ func RunCreate(f *cmdutil.Factory, out io.Writer, filenames util.StringList) err
...
+			`, makePortsString(obj.Spec.Ports))
+			out.Write([]byte(msg))
+		}
+	}
+}
+
+func makePortsString(ports []api.ServicePort) string {
+	pieces := make([]string, len(ports))
+	for ix := range ports {
+		pieces[ix] = fmt.Sprintf("%s:%d", ports[ix].Protocol, ports[ix].Port)
```
Comment (thockin) [inline pkg/kubectl/cmd/create.go]: do you need to ToLower() the protocol?

File: pkg/kubectl/cmd/create.go [hunk 3]
```diff
@@ -102,6 +105,7 @@ func RunCreate(f *cmdutil.Factory, out io.Writer, filenames util.StringList) err
 		}
 		count++
 		info.Refresh(obj, true)
+		printObjectSpecificMessage(info.Object, out)
```
Comment (thockin) [inline pkg/kubectl/cmd/create.go]: what about Updates?

Comment (brendandburns) [inline pkg/kubectl/cmd/create.go]: done.

File: pkg/kubectl/cmd/create.go (hunk 2 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/create.go]: done.

File: pkg/kubectl/cmd/create.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/create.go]: done and done.

**Comment by brendandburns** [comment]
Comments addressed.  ptal.

Thanks!
--brendan

File: docs/services-firewalls.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Services and Firewalls
+
+Many cloud providers (e.g. Google Compute Engine) define firewalls that help keep prevent inadvertent
```
Comment (thockin) [inline docs/services-firewalls.md]: s/keep/

File: pkg/kubectl/cmd/create.go [hunk 5]
```diff
@@ -113,3 +117,36 @@ func RunCreate(f *cmdutil.Factory, out io.Writer, filenames util.StringList) err
...
+
+func printObjectSpecificMessage(obj runtime.Object, out io.Writer) {
+	switch obj := obj.(type) {
+	case *api.Service:
+		if obj.Spec.Type == api.ServiceTypeLoadBalancer {
+			msg := fmt.Sprintf(`
+			An external load-balanced service was created.  On many platforms (e.g. Google Compute Engine),
+			you will also need to explicitly open a Firewall rule for the service port(s) (%s) to serve traffic.
+
+			See https://github.com/GoogleCloudPlatform/kubernetes/tree/master/docs/services-firewall.md for more details.
```
Comment (thockin) [inline pkg/kubectl/cmd/create.go]: Can we use the docs.k8s.io short URL here?

**Comment by thockin** [comment]
nits then LGTM

**Comment by quinton-hoole** [comment]
Merging for now to prevent unnecessary rebase hassle.  @brendanbburns, please fix the comment nits in a separate PR.

File: pkg/kubectl/cmd/create.go [hunk 6]
```diff
@@ -113,3 +117,36 @@ func RunCreate(f *cmdutil.Factory, out io.Writer, filenames util.StringList) err
...
+			out.Write([]byte(msg))
+		}
+		if obj.Spec.Type == api.ServiceTypeNodePort {
+			msg := fmt.Sprintf(`
+				You have exposed your service on an external port on all nodes in your cluster.
+				If you want to expose this service to the external internet, you may need to set up
+				firewall rules for the service port(s) (%s) to serve traffic.
+				
+				See https://github.com/GoogleCloudPlatform/kubernetes/tree/master/docs/services-firewall.md for more details.
+				`, makePortsString(obj.Spec.Ports))
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/create.go]: It seems a user needs a firewall rule to allow traffic to the nodePort, not the service port, doesn't it? @thockin @brendandburns

**Pull request merged**